alarm_active_table.cpp

```diff
@@ -0,0 +1,453 @@
+      delete[](model_pointer);
+
+    }
+    break;
+    
+    case COLUMN_ALARMACTIVESPECIFICPOINTER:
+    {
+      // This would point to the model-specific active alarm list
+      // for this alarm. We do not support model-specific Alarm-MIB
+      // and hence this will always be the 0.0 OID
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) zero_dot_zero_oid,
+                               sizeof(zero_dot_zero_oid));
+    }
+    break;
+    
+    
+    default: /** We shouldn't get here */
+    {
+      TRC_ERROR("unknown column: alarmActiveTable_get_value");
+      return SNMP_ERR_GENERR;
+    }
+  }
+  return SNMP_ERR_NOERROR;
+}
+
+/*************************************************************
+ *
+ * Find the current time and use it to fill the d_time struct
+ * which is formatted to be compliant with RFC 3877
+ */
+alarmActiveTable_SNMPDateTime alarm_time_issued(void)
+{
+  alarmActiveTable_SNMPDateTime d_time;
+  struct timespec ts;
+  struct tm *timing = new tm;
+  time_t rawtime;
+  int UTC_offset;
+
+  clock_gettime(CLOCK_REALTIME, &ts);
+  rawtime = ts.tv_sec;
+  // localtime_r converts calendar time to a broken down representation
+  // to be stored in the stucture timing
+  localtime_r(&rawtime, timing);
+  // This updates the variable timezone declared within time.h which
+  // tells us how offset (in seconds) from UTC we are.
+  localtime(&rawtime);
+  UTC_offset = abs(timezone) / 3600;
+  // htons converts the integer from host byte order to network
+  // byte order as specified in RFC 2579.
+  d_time.year = htons(1900 + timing->tm_year);
+  d_time.month = 1 + timing->tm_mon;
+  d_time.day = timing->tm_mday;
+  d_time.hour = timing->tm_hour;
+  d_time.minute = timing->tm_min;
+  d_time.second = timing->tm_sec;
+  d_time.decisecond = 0;
+  d_time.direction = (timezone > 0) ? '+' : '-';
+  d_time.timegeozone = UTC_offset;
+  d_time.mintimezone = (abs(timezone) % 3600) / 60;
+  delete(timing);
+  return d_time;
+}
+
+/***********************************************************
+ *
+ * Logic tests to determine how to handle each alarm that
+ * is raised. This ensures each alarm has at most one entry
+ * in Alarm Active Table.
+ */
+void alarmActiveTable_trap_handler(AlarmTableDef& def)
+{
+  alarmActiveTable_context* existing_row = alarm_indexes_to_rows[def.alarm_index()];
+  // 1) New alarm
+  if (existing_row == NULL && def.severity() != AlarmDef::CLEARED)
+  {
+  alarmActiveTable_create_row(def); 
+  }
+  // 2) Existing alarm with different severity
+  else if (existing_row != NULL && def.severity() != AlarmDef::CLEARED && def.severity() != existing_row->_alarm_table_def->severity())
+  {
+  alarmActiveTable_delete_row(def);
+  alarmActiveTable_create_row(def);
+  }
+  // 3) Existing alarm with same severity
+  else if (existing_row != NULL && def.severity() != AlarmDef::CLEARED && def.severity() == existing_row->_alarm_table_def->severity())
+  {
+  return;
```

Comment:
I probably wouldn't return at this point - as a general rule, it's good to have as few return statements in a function as possible (so that if you want to free memory, or log a message, before returning, you have fewer places to do it and less chance of missing one). I'd either have a comment just saying `// No-op`, or exclude this case from the if statement entirely (so it only had the three cases where we need to do something - raising, updating or clearing an alarm).